hermeto/core/package_managers/yarn/resolver.py

```diff
@@ -455,18 +459,28 @@ def _cache_path_as_rooted(self, cache_path: str) -> RootedPath:
 
     def _get_path_patch_url(self, patch_locator: PatchLocator, patch_path: Path) -> str:
         """Return a PURL-style VCS URL qualifier with subpath for a Patch."""
-        if patch_locator.locator is None:
-            raise UnsupportedFeature(
-                f"{patch_locator} is missing an associated workspace locator "
-                "and {APP_NAME} expects all non-builtin yarn patches to have one"
-            )
+        patch_str = str(patch_path)
+        pp_root = self._project.source_dir.root
+        pp_join = self._project.source_dir.join_within_root
+        project_relative_prefix = "~/"
+
+        # Note that RootedPath can raise PathOutsideRoot
+        if patch_path.is_absolute():
+            normalized = pp_join(patch_path)
+        elif patch_str.startswith(project_relative_prefix):
+            project_relative_path = Path(patch_str.removeprefix(project_relative_prefix))
+            normalized = pp_join(project_relative_path)
+        else:
+            locator = patch_locator.locator
 
-        project_path = self._project.source_dir
-        workspace_path = patch_locator.locator.relpath
-        normalized = self._project.source_dir.join_within_root(workspace_path, patch_path)
-        repo_url = get_repo_id(project_path.root).as_vcs_url_qualifier()
-        subpath_from_root = str(normalized.subpath_from_root)
+            # handle locator = None case
```

Comment:
Another nitpick: if locator can actually be `None` it is better (as per PEP8) to be explicit about checking if it `is None`. That would also make the comment unnecessary.